kernel/sched.c

```diff
@@ -129,35 +161,43 @@ static struct k_thread *next_up(void)
 	/* Under SMP, the "cache" mechanism for selecting the next
 	 * thread doesn't work, so we have more work to do to test
 	 * _current against the best choice from the queue.
+	 *
+	 * Subtle note on "queued": in SMP mode, _current does not
+	 * live in the queue, so this isn't exactly the same thing as
+	 * "ready", it means "is _current already added back to the
+	 * queue such that we don't want to re-add it".
 	 */
-	int active = !_is_thread_prevented_from_running(_current);
 	int queued = _is_thread_queued(_current);
+	int active = !_is_thread_prevented_from_running(_current);
 
+	/* Choose the best thread that is not current */
 	struct k_thread *th = _priq_run_best(&_kernel.ready_q.runq);
-
-	/* Idle thread if nothing else */
 	if (!th) {
 		th = _current_cpu->idle_thread;
 	}
 
-	/* Stay with current unless it's already been put back in the
-	 * queue and something better is available (c.f. timeslicing,
-	 * yield)
-	 */
-	if (active && !queued && !_is_t1_higher_prio_than_t2(th, _current)
-	    && !is_metairq(th))	{
-		th = _current;
+	if (active) {
+		if (!queued &&
+		    !_is_t1_higher_prio_than_t2(th, _current)) {
+			th = _current;
+		}
+
```

Comment:
Maybe `if ((!queued && prio(t1)>prio(t2)) || !should_preempt(...))`, to avoid calling `should_preempt()` in cases where `!queued && prio(t1)>prio(t2)`?